Vonk/vonk 326 create bundles (#51)
simplifhir/simplifier-docs: vonk/reference/programming_api/bundles.rst

.. _vonk_reference_api_bundles:

Constructing bundles
====================

In a Plugin or Facade you may need to construct a bundle from a set of resources, e.g. a SearchResult (see :ref:`here <vonk_reference_api_isearchrepository>`). There are two ways of doing this: with the Bundle POCO or with SourceNodes.

Bundle POCO
-----------

> This is fairly straightforward. Create a new Bundle object, and fill its properties, iterating over the set of resources that you have. The code looks like this:

Reviewer: I thought Ewout wanted POCOs deprecated at some point? If that's the case, maybe we shouldn't promote it as the first choice?

Author: POCOs are not deprecated any time soon. But internally for our tooling - that often has to support multiple FHIR versions - we'd rather not use them. If you make a plugin that is only used for R4 - by all means use R4 POCO's.

.. code-block:: csharp

   using Hl7.Fhir.Model; //Either from Hl7.Fhir.Core.Stu3 or Hl7.Fhir.Core.R4

   ...

      var searchResult = await searchRepository.Search(args, options)

>       var bundle = new ResourceKey(resourceType, resourceId);

Reviewer 2: For what do we need the ResourceKey? Bundle needs also to be initialised with new Bundle.
I would also add a comment that Bundle.type is mandatory.

      foreach (var resource in searchResult)
      {
         bundle.Entry.Add
         {
            new BundleEntryComponent
            {
               Resource = resource.ToPoco<Resource>();
               // fill in any other details like Request or Response.
            }
         }
      }

      //fill in details of the bundle as a whole, like Meta or Identifier.

> The limitation of this is that you are bound to either STU3 or R4. That implies that you cannot include :ref:`feature_customresources` in the bundle.

Reviewer: ```suggestion
The limitation of this is that you are bound to either STU3 or R4, and that also implies that you cannot include :ref:`feature_customresources` in the bundle.
```

Bundle from SourceNodes
-----------------------

> ISourceNode is from Hl7.Fhir.ElementModel and not tied to a specific FHIR version. It can represent just about any tree. Vonk can serialize ISourceNode provided that the right StructureDefinition is available in the Administration API. For Bundles that is by default the case. That means you can construct a Bundle using ISourceNode. 

Reviewer: ```suggestion
ISourceNode is from Hl7.Fhir.ElementModel and not tied to a specific FHIR version, and Vonk can serialize ISourceNode provided that the right StructureDefinition is available in the Administration API - which is the case for Bundle by default.
```

You start by creating the Bundle itself using the class SourceNode that allows for construction of ISourceNode nodes. 

.. code-block:: csharp

   using Hl7.Fhir.ElementModel;

   ...

      var bundleNode = SourceNode.Resource("Bundle", "Bundle", SourceNode.Valued("type", "document")); 
      
      //choose type as one of the bundle types from the spec, see http://hl7.org/fhir/R4/bundle-definitions.html#Bundle.type

Then you can add elements to the bundle itself that are not in the entries of the bundle. Like an identifier:

.. code-block:: csharp

      var identifier = SourceNode.Node("identifier");
      identifier.Add(SourceNode.Valued("system", "urn:ietf:rfc:3986"));
      identifier.Add(SourceNode.Valued("value", Guid.NewGuid().ToString()));
      bundleNode.Add(identifier);

Then you can turn this into the helper class ``GenericBundle`` that provides several helper methods on an ISourceNode that is known to be a Bundle.

.. code-block:: csharp

>       var documentBundle = GenericBundle.FromBundle(bundleResourceNode);

Reviewer 3: bundleResourceNode should probably be bundleNode.
